crates/bevy_ecs/src/query/state.rs

```diff
@@ -2030,4 +2045,32 @@ mod tests {
         let query_2 = QueryState::<&B, Without<C>>::new(&mut world);
         let _: QueryState<Entity, Changed<C>> = query_1.join_filtered(&world, &query_2);
     }
+
+    #[test]
+    fn query_respects_default_filters() {
+        let mut world = World::new();
+        world.spawn((A(0), B(0)));
+        world.spawn((B(0), C(0)));
+        world.spawn(C(0));
+
+        let b = world.component_id::<B>().unwrap();
+        let c = world.component_id::<C>().unwrap();
+
+        let mut default_filters = DefaultQueryFilters::default();
+        default_filters.with(b);
+        default_filters.without(c);
+        world.insert_resource(default_filters);
+
+        // With<B>, Without<C> only matches the first entity
+        let mut query = QueryState::<()>::new(&mut world);
+        assert_eq!(1, query.iter(&world).count());
```

Comment:
I worry about what would happen if the order of setting up `DefaultQueryFilters` and the `QueryState` is inverted. Would the `QueryState` ignore the filters? That feels like a huge footgun.

Reply:
This doesn't solve the issue when a plugin creates a `QueryState`/`SystemState` while building the plugin, as it won't use the "full"/"final" `DefaultQueryFilters`. Note that such plugin may not even use query filters itself, so the author might not have read the documentation of `DefaultQueryFilters`, so it would have to be mentioned in the documentation of `QueryState`/`SystemState`.

---

Also, since adding default filters after plugin construction is basically guaranteed to produce incorrect behaviour I would propose to make that panic if it happens. Better failing fast than breaking things :)

The caveats should also be mentioned in `World::set_default_with_filter` and `World::set_default_without_filter` since those are the methods that people will use most often. Or at least mention the fact that there are caveats and to look at the documentation for `DefaultQueryFilters` to see them (currently it links to it but doesn't mention the fact there are caveats).